docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,457 @@
+  parameters:
+    keyStores:
+      - inline-key
+```
+- Bob attempts to deploy a pod from the vetted image that has 2 cosign signatures, one of which is signed with key in CertificateStore `inline-key`. Pod is verified and created successfully.
+- Bob attempts to deploy a pod from an image that has cosign signature(s) signed with a different key in `inline-key`. Pod FAILS verification and blocked.
+
+#### 2 Signatures 2 Keys: both keys must be used
+
+Bob has a container image that is produced from a build pipeline and tested via a testing pipeline. After each pipeline, the image is signed with cosign using a SEPARATE key. Now the image has 2 cosign signatures. Bob trusts both keys and requires BOTH keys to be used. Bob wants to ensure all container images entering his K8s cluster are verified to have a valid cosign signature from his build AND test pipeline.
+
+- Bob installs Ratify on the cluster
+- Bob applies a new inline `CertificateStore` CR onto the cluster, `inline-key-buid` & `inline-key-test`
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: CertificateStore
+metadata:
+  name: inline-key-build
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  provider: inline
+  parameters:
+    type: key
+    value: |
+    ---------- BEGIN RSA KEY ------------
+    ******
+    ---------- END RSA KEY ------------
+    
+```
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: CertificateStore
+metadata:
+  name: inline-key-test
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  provider: inline
+  parameters:
+    type: key
+    value: |
+    ---------- BEGIN RSA KEY ------------
+    ******
+    ---------- END RSA KEY ------------
+    
+```
+
+- Bob applies the cosign verifier CR
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: Verifier
+metadata:
+  name: verifier-cosign
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  name: cosign
+  artifactTypes: application/vnd.dev.cosign.artifact.sig.v1+json
+  parameters:
+    keyStores:
+      - inline-key-build
+      - inline-key-test
+    keyEnforcement: all
+```
+- Bob attempts to deploy a pod from the vetted image that has 1 cosign signature, which is signed with key in CertificateStore `inline-key-build`. Pod FAILS verification and blocked.
+- Bob attempts to deploy a pod from the vetted image that has 2 cosign signatures, which is signed with keys in CertificateStore `inline-key-build` & `inline-key-test`. Pod passes verification and is created.
+
+> **NOTE**: Based on Cosign's own verification implementation, there does not seem to be a way to enforce that ALL signatures found are valid. This behavior is unchartered as far as we can tell and would require Ratify to determine if we want to support this directly. For that reason, the strict ALL signature scenario is not yet included.
```

Comment:
I feel it's a valid case since we already support this scenario in notation verifier. The question is that if we want to have single cosign verifier to enforce it or have different cosign verifiers with different keys.

Reply:
very true. we need to discuss this. I will bring this up as topic to discuss